pinot-query-planner/src/main/java/org/apache/pinot/query/routing/WorkerManager.java

```diff
@@ -183,4 +197,34 @@ private RoutingTable getRoutingTable(String tableName, TableType tableType, long
     return _routingManager.getRoutingTable(
         CalciteSqlCompiler.compileToBrokerRequest("SELECT * FROM " + tableNameWithType), requestId);
   }
+
+  private boolean isLeafStage(StageMetadata stageMetadata) {
+    return stageMetadata.getScannedTables().size() == 1;
```

Comment:
I don't think this is a reliable indicator. May be true for now but unlikely to give the right answer as functionality evolves. 

Consider a standard GROUP BY query run in multi stage engine where the intermediary stage also processes the same 1 table after hash exchange. Even there the stageMetadata will indicate single table ?

Also, does `getScannedTables()` actually refer to true scan of the tables -- which will only happen on leaf stage. Since the non leaf stages will always work off blocks and thus `getScannedTables()` will return empty, may be one way to indicate leaf v/s non-leaf is if `getScannedTables().size() > 0`

Reply:
We can also have stages terminate in a value node in which case there wouldn't be any scanned tables.

Reply:
Agreed that we need a better way to figure out whether it is a leaf stage or not. 

> We can also have stages terminate in a value node in which case there wouldn't be any scanned tables.

True, we actually treat these stages as intermediate stages today.


Note that this PR only moved this check into a function. I'm leaving a TODO to address this.